apps/emqx_auth_mnesia/src/emqx_authz_mnesia.erl

```diff
@@ -124,75 +148,110 @@ init_tables() ->
+            (?WHO_NS(Ns, _) = Key) when Ns == Namespace ->
+                ok = mria:dirty_delete(?ACL_NS_TABLE, Key);
+            (_Key) ->
+                ok
+        end,
+        mnesia:dirty_all_keys(?ACL_NS_TABLE)
     ).
 
 %% @doc Get one record
--spec get_rules(who()) -> {ok, rules()} | not_found.
-get_rules({username, Username}) ->
-    do_get_rules({?ACL_TABLE_USERNAME, Username});
-get_rules({clientid, Clientid}) ->
-    do_get_rules({?ACL_TABLE_CLIENTID, Clientid});
-get_rules(all) ->
-    do_get_rules(?ACL_TABLE_ALL).
+-spec get_rules(maybe_namespace(), who()) -> {ok, rules()} | not_found.
+get_rules(Namespace, {username, Username}) ->
+    do_get_rules(Namespace, {?ACL_TABLE_USERNAME, Username});
+get_rules(Namespace, {clientid, Clientid}) ->
+    do_get_rules(Namespace, {?ACL_TABLE_CLIENTID, Clientid});
+get_rules(Namespace, all) ->
+    do_get_rules(Namespace, ?ACL_TABLE_ALL).
 
 %% @doc Delete one record
--spec delete_rules(who()) -> ok.
-delete_rules({username, Username}) ->
-    mria:dirty_delete(?ACL_TABLE, {?ACL_TABLE_USERNAME, Username});
-delete_rules({clientid, Clientid}) ->
-    mria:dirty_delete(?ACL_TABLE, {?ACL_TABLE_CLIENTID, Clientid});
-delete_rules(all) ->
-    mria:dirty_delete(?ACL_TABLE, ?ACL_TABLE_ALL).
-
--spec list_username_rules() -> ets:match_spec().
-list_username_rules() ->
+-spec delete_rules(maybe_namespace(), who()) -> ok.
+delete_rules(Namespace, {username, Username}) ->
+    do_delete_one(Namespace, {?ACL_TABLE_USERNAME, Username});
+delete_rules(Namespace, {clientid, Clientid}) ->
+    do_delete_one(Namespace, {?ACL_TABLE_CLIENTID, Clientid});
+delete_rules(Namespace, all) ->
+    do_delete_one(Namespace, ?ACL_TABLE_ALL).
+
+-spec list_username_rules(maybe_namespace()) -> ets:match_spec().
+list_username_rules(?global_ns) ->
+    ets:fun2ms(
+        fun(#?ACL_TABLE{who = {?ACL_TABLE_USERNAME, Username}, rules = Rules}) ->
+            [{username, Username}, {rules, Rules}]
+        end
+    );
+list_username_rules(Namespace) when is_binary(Namespace) ->
     ets:fun2ms(
-        fun(#emqx_acl{who = {?ACL_TABLE_USERNAME, Username}, rules = Rules}) ->
+        fun(#?ACL_NS_TABLE{who = ?WHO_NS(Ns, {?ACL_TABLE_USERNAME, Username}), rules = Rules}) when
+            Ns == Namespace
+        ->
             [{username, Username}, {rules, Rules}]
         end
     ).
 
--spec list_clientid_rules() -> ets:match_spec().
-list_clientid_rules() ->
+-spec list_clientid_rules(maybe_namespace()) -> ets:match_spec().
+list_clientid_rules(?global_ns) ->
     ets:fun2ms(
-        fun(#emqx_acl{who = {?ACL_TABLE_CLIENTID, Clientid}, rules = Rules}) ->
+        fun(#?ACL_TABLE{who = {?ACL_TABLE_CLIENTID, Clientid}, rules = Rules}) ->
+            [{clientid, Clientid}, {rules, Rules}]
+        end
+    );
+list_clientid_rules(Namespace) when is_binary(Namespace) ->
+    ets:fun2ms(
+        fun(#?ACL_NS_TABLE{who = ?WHO_NS(Ns, {?ACL_TABLE_CLIENTID, Clientid}), rules = Rules}) when
+            Ns == Namespace
```

Comment:
As far as I understand, such `ets:fun2ms` will produce a match spec with Ns constraint in Guard, not in MatchHead, which will result into full scans.
We probably should compose the matchspec manually to take advantage of ets ordering.